### ISSUE TITLE
Nested maps not hoisted out of the internal list of maps representation.

### ISSUE THREAD

**Issue by jbardin** [hashicorp/terraform#11115]
When trying to create a nested map the inner map (value of the outer map) is contained in a list.  This does not happen if I use the map() though.

main.tf
```
resource "null_resource" "n" {
  triggers {}
}

output "nested_map" {
  value = {
    foo = {
      a = "value1"
      b = "value2"
    }
    bar = {
      a = "value3"
      b = "value4"
    }
  }
}

output "nested_map_function" {
  value = "${
    map(
      "foo", map("a", "value1", "b", "value2"),
      "bar", map("a", "value3", "b", "value4")
    )
  }"
}
```

outputs:
```
nested_map = {
  bar = [map[b:value4 a:value3]]
  foo = [map[b:value2 a:value1]]
}
nested_map_function = {
  bar = map[a:value3 b:value4]
  foo = map[a:value1 b:value2]
}
```

outputs JSON:
```
{
    "nested_map": {
        "sensitive": false,
        "type": "map",
        "value": {
            "bar": [
                {
                    "a": "value3",
                    "b": "value4"
                }
            ],
            "foo": [
                {
                    "a": "value1",
                    "b": "value2"
                }
            ]
        }
    },
    "nested_map_function": {
        "sensitive": false,
        "type": "map",
        "value": {
            "bar": {
                "a": "value3",
                "b": "value4"
            },
            "foo": {
                "a": "value1",
                "b": "value2"
            }
        }
    }
}
```

**Comment by mitchellh** [comment]
We can add this to the list of things. Nested maps cause pain in many more places, too.

**Comment by tmclaugh** [comment]
I've been meaning to file an additional bug but have been distracted.  The following throws are parser error.  Want a new bug or know issue?

```
resource "aws_autoscaling_group" "asg" {
  vpc_zone_identifier   = [
    "${data.terraform_remote_state.infrastructure.vpc.subnet_ids[var.vpc_name][var.subnet_type]}"
  ]
<snip>
```

```
Error loading Terraform: Error downloading modules: module central_registration_nodejs: Error loading .terraform/modules/f6da1e738ab0286d0d4b1dd6e259b3f5/main.tf: Error reading config for aws_autoscaling_group[asg]: parse error at 1:74: expected "}" but found "["
```

**Comment by apparentlymart** [comment]
Verified fixed in v0.12.0-alpha2:

```
Outputs:

nested_map = {
  "bar" = {
    "a" = "value3"
    "b" = "value4"
  }
  "foo" = {
    "a" = "value1"
    "b" = "value2"
  }
}
nested_map_function = {
  "bar" = {
    "a" = "value3"
    "b" = "value4"
  }
  "foo" = {
    "a" = "value1"
    "b" = "value2"
  }
}
```
